src/multi_vector_simulator/E3_indicator_calculation.py

```diff
@@ -502,11 +580,181 @@ def equation_degree_of_sector_coupling(
+        # Get source connected to the specific DSO in question
+        scalars_matrix = dict_values[KPI][KPI_SCALAR_MATRIX]
+        # load total flow into the dso sink
+        dso_sink = scalars_matrix.loc[
+            scalars_matrix["label"]
+            == dict_values[ENERGY_PROVIDERS][  # todo: IS this referece always correct??
+                dso
+            ][CONNECTED_FEEDIN_SINK]
+            + AUTO_SINK
+        ]
+        dso_feedin = dso_sink["total_flow"][0]
+        # sum up all dso flows
+        total_feedin = total_feedin + dso_feedin
+    # save total feedin into KPI_SCALARS_DICT
+    dict_values[KPI][KPI_SCALARS_DICT].update({TOTAL_FEEDIN: total_feedin})
+
+    # calculate onsite energy fraction
+    onsite_energy_fraction = equation_onsite_energy_fraction(
+        total_generation, total_feedin
+    )
+
+    # save KPI  onsite energy fraction into KPI_SCALARS_DICT
+    dict_values[KPI][KPI_SCALARS_DICT].update(
+        {ONSITE_ENERGY_FRACTION: onsite_energy_fraction}
+    )
+    logging.debug(
+        f"Calculated the {ONSITE_ENERGY_FRACTION}: {round(onsite_energy_fraction)}"
+    )
+    logging.info(f"Calculated the {ONSITE_ENERGY_FRACTION} of the LES.")
+
+    return
+
+
+def equation_onsite_energy_fraction(total_generation, total_feedin):
+    """
+    Calculates onsite energy fraction (OEF), i.e. self-consumption.
+
+    OEF describes the fraction of all locally generated energy that is consumed
+    by the system itself.
+
+        Parameters
+        ----------
+        total_generation: float
+            Energy equivalent of total generation flows
+        total_feedin: float
+            Total feed into the grid
+
+        Returns
+        -------
+            float
+                Onsite energy fraction.
+
+        .. math::
+                OEF &=\frac{\sum_{i} {E_{generation} (i) \cdot w_i} - E_{gridfeedin}(i) \cdot w_i}{\sum_{i} {E_{generation} (i) \cdot w_i}}
+
+                &OEF \epsilon \text{[0,1]}
+
+        """
+    onsite_energy_fraction = (total_generation - total_feedin) / total_generation
+
     return onsite_energy_fraction
 
 
-def equation_onsite_energy_matching():
+def add_onsite_energy_matching(dict_values):
+    """
+    Determines onsite energy matching (OEM), i.e. self-sufficiency, and adds KPI to dict_values
+
+    Parameters
+    ----------
+    dict_values: dict
+        dict with all project information and results
+        after applying total_renewable_and_non_renewable_energy_origin and
+        total_demand_and_excess_each_sector and
+        add_onsite_energy_fraction
+    
+    Returns
+    -------
+    None
+        updated dict_values with onsite energy matching KPI
+
+    Tested with
+    - test_equation_onsite_energy_matching()
+    """
+
+    total_generation = (
+        dict_values[KPI][KPI_SCALARS_DICT][TOTAL_RENEWABLE_GENERATION_IN_LES]
+        + dict_values[KPI][KPI_SCALARS_DICT][TOTAL_NON_RENEWABLE_GENERATION_IN_LES]
+    )
```

Comment:
If you calculate this for each function again and again, wouldnt it be better to add a KPI for this in the function that results in `TOTAL_NON_RENEWABLE_GENERATION_IN_LES `and `TOTAL_NON_RENEWABLE_GENERATION_IN_LES`: `TOTAL_GENERATION_IN_LES`?

Reply:
done